### PR TITLE
AutoYaST: remove code duplication when creating/reusing partitions

### PR DESCRIPTION
**Pull request by ancorgs**
## Problem

The AutoYaST code to create partitions in disks, in MD RAIDs and in bcache devices was too similar. Specially the `MdCreator` and `BcacheCreator` classes were almost identical.

See https://trello.com/c/wmMCpUkq/985-5-autoyast-remove-duplication-between-raid-and-bcache-code

## Solution

A new mixin and a new class created to kill the duplication
  * The `Proposal::AutoinstPartitionSize` mixin encapsulates the logic for calculating mix/max/desired sizes for the partitions, based on AutoYaST size specifications
   * The `Proposal::AutoinstPartitioner` class encapsulates the creation and reuse of partitions in all scopes (disks, MD, bcache).

## Testing

- All unit tests moved to the right location or replaced with a new one

## Note for reviewers

The first two commits are just to rename the `MdCreator` and `BcacheCreator` classes, since they actually contained a lot of AutoYaST specific code. The other two commits are better reviewed one by one.

### CODE REVIEW

File: src/lib/y2storage/proposal/autoinst_bcache_creator.rb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        partitioner.process_device(bcache, planned_bcache)
+      end
+
+    private
+
+      # @param planned_bcache  [Planned::Bcache] bcache device
+      # @param backing_devname [String] Backing device name
+      # @param caching_devname [String] Caching device name
+      # @return [Planned::Bcache]
+      def create_bcache_device(devicegraph, planned_bcache, backing_devname, caching_devname)
```
Comment (imobachgs) [inline src/lib/y2storage/proposal/autoinst_bcache_creator.rb]: Please, add missing documentation for `devicegraph` argument.

File: src/lib/y2storage/proposal/autoinst_partition_size.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      # partitions will be all created in such device. If not, a devicegraph
+      # must be provided and the device for each partition will be searched in
+      # the devicegraph using the corresponding #{Planned::Partition#disk}
+      # attribute.
+      #
+      # @raise [ArgumentError] if no device or devicegraph is provided.
+      #
+      # @param planned_partitions [Array<Planned::Partition>] List of planned partitions
+      # @param device [Partitionable, nil] concrete device to partition, if any
+      # @param devicegraph [Partitionable] devicegraph to look up the devices,
```
Comment (imobachgs) [inline src/lib/y2storage/proposal/autoinst_partition_size.rb]: From the previous description, I understood that `devicegraph` could be `nil`.

File: src/lib/y2storage/proposal/autoinst_partition_size.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      # Returns a list of planned partitions adjusting the size
+      #
+      # All partitions which sizes are specified as percentage will get their minimal and maximal
+      # sizes adjusted.
+      #
+      # If a device is provided as argument, it is considered that the
+      # partitions will be all created in such device. If not, a devicegraph
+      # must be provided and the device for each partition will be searched in
+      # the devicegraph using the corresponding #{Planned::Partition#disk}
+      # attribute.
```
Comment (imobachgs) [inline src/lib/y2storage/proposal/autoinst_partition_size.rb]: np: From my POV, having two named arguments which are exclusive is a little bit confusing.

File: src/lib/y2storage/proposal/autoinst_partition_size.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+          new_part
+        end
+      end
+
+      # Return a list of new planned devices with flexible limits
+      #
+      # The min_size is removed and a proportional weight is set for every device.
+      #
+      # @return [Array<Planned::Partition>]
+      def flexible_partitions(devices)
```
Comment (imobachgs) [inline src/lib/y2storage/proposal/autoinst_partition_size.rb]: Please, document `devices` argument.

File: src/lib/y2storage/proposal/autoinst_partitioner.rb [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+      # Constructor
+      #
+      # @param devicegraph [Devicegraph] see {#devicegraph}
+      def initialize(devicegraph)
+        @devicegraph = devicegraph
+      end
+
+      # Reuses partitions in the target devicegraph
+      #
```
Comment (imobachgs) [inline src/lib/y2storage/proposal/autoinst_partitioner.rb]: Please, add a note to make clear that the `devicegraph` is modified (given that's the case).

File: src/lib/y2storage/proposal/autoinst_partitioner.rb [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+      # @param reused_parts  [Array<Planned::Partition>] Partitions to reuse
+      def reuse_partitions(reused_parts)
+        reuse_partitions_in_devicegraph(reused_parts, devicegraph)
+      end
+
+      # Reuses existing partitions of the given planned device
+      #
+      # @note This method does not modify the original devicegraph but returns
```
Comment (imobachgs) [inline src/lib/y2storage/proposal/autoinst_partitioner.rb]: :+1:

File: src/lib/y2storage/proposal/autoinst_partition_size.rb (hunk 3 above)
Comment (ancorgs) [reply src/lib/y2storage/proposal/autoinst_partition_size.rb]: I agree. But being just a mixin for kind of internal usage, I considered it acceptable.

**Pull request merged**